clouddriver-aws/src/main/java/com/netflix/spinnaker/clouddriver/aws/deploy/asg/AutoScalingWorker.java

```diff
@@ -0,0 +1,345 @@
+   *   <li>Look up an ancestor ASG based on Netflix naming conventions, and bring its security
+   *       groups to the new ASG;
+   *   <li>Retrieve user data from all available {@link
+   *       com.netflix.spinnaker.clouddriver.aws.deploy.userdata.UserDataProvider}s;
+   *   <li>Create the ASG's Launch Configuration or Launch Template with User Data and Security
+   *       Groups;
+   *   <li>Create a new ASG in the subnets found from the optionally supplied subnetType.
+   * </ol>
+   *
+   * @return the name of the newly deployed ASG
+   */
+  public String deploy(AsgConfiguration asgConfig) {
+    getTask().updateStatus(AWS_PHASE, "Beginning Amazon deployment.");
+
+    if (asgConfig.startDisabled != null && asgConfig.startDisabled) {
+      asgConfig.suspendedProcesses.addAll(
+          AutoScalingProcessType.getDisableProcesses().stream()
+              .map(AutoScalingProcessType::name)
+              .collect(Collectors.toList()));
+    }
+
+    getTask().updateStatus(AWS_PHASE, "Beginning ASG deployment.");
+
+    AWSServerGroupNameResolver awsServerGroupNameResolver =
+        regionScopedProvider.getAWSServerGroupNameResolver();
+    String asgName;
+    if (asgConfig.sequence != null) {
+      asgName =
+          awsServerGroupNameResolver.generateServerGroupName(
+              asgConfig.application,
+              asgConfig.stack,
+              asgConfig.freeFormDetails,
+              asgConfig.sequence,
+              false);
+    } else {
+      asgName =
+          awsServerGroupNameResolver.resolveNextServerGroupName(
+              asgConfig.application,
+              asgConfig.stack,
+              asgConfig.freeFormDetails,
+              asgConfig.ignoreSequence);
+    }
+
+    AsgBuilder asgBuilder;
+    if (shouldSetLaunchTemplate(asgConfig)) {
+      // process the IPv6 setting conditionally
+      if (asgConfig.getAssociateIPv6Address() == null) {
+        String asgConfigEnv = asgConfig.getCredentials().getEnvironment();
+        Boolean autoEnableIPv6 =
+            dynamicConfigService.getConfig(
+                Boolean.class, "aws.features.launch-templates.ipv6." + asgConfigEnv, false);
+        asgConfig.setAssociateIPv6Address(autoEnableIPv6);
+      }
+
+      if (asgConfig.shouldUseMixedInstancesPolicy()) {
+        asgBuilder = regionScopedProvider.getAsgBuilderForMixedInstancesPolicy();
+      } else {
+        asgBuilder = regionScopedProvider.getAsgBuilderForLaunchTemplate();
+      }
+    } else {
+      asgBuilder = regionScopedProvider.getAsgBuilderForLaunchConfiguration();
+    }
+
+    return asgBuilder.build(getTask(), AWS_PHASE, asgName, asgConfig);
+  }
+  /** This is used to gradually roll out launch template. */
+  private boolean shouldSetLaunchTemplate(final AsgConfiguration asgConfig) {
+    // Request level flag that forces launch configurations.
+    if (asgConfig.setLaunchTemplate != null && !asgConfig.setLaunchTemplate) {
```

Comment:
Wouldn't the groovy check:
```
if (!asgConfig.setLaunchTemplate) {
```
be true if setLaunchTemplate is null?  So then the code here would be:
```suggestion
    if (asgConfig.setLaunchTemplate == null || !asgConfig.setLaunchTemplate) {
```
?

Reply:
You are right :) Thanks for catching this.